gobblin-runtime/src/main/java/org/apache/gobblin/runtime/StreamModelTaskRunner.java

```diff
@@ -155,13 +156,17 @@ protected void run() throws Exception {
     Thread thread = new Thread(() -> connectableStream.connect());
     thread.setName(this.getClass().getSimpleName());
     //Log uncaught exceptions (e.g.OOMEs) to prevent threads from dying silently
-    thread.setUncaughtExceptionHandler(new LoggingUncaughtExceptionHandler(Optional.absent()));
+    LoggingUncaughtExceptionHandler exceptionHandler = new LoggingUncaughtExceptionHandler(Optional.absent());
+    thread.setUncaughtExceptionHandler(exceptionHandler);
     thread.start();
-
-    if (!ExponentialBackoff.awaitCondition().callable(() -> this.forks.keySet().stream().map(Optional::get).allMatch(Fork::isDone)).
+    if (!ExponentialBackoff.awaitCondition().callable(() ->
+        this.forks.keySet().stream().map(Optional::get).allMatch(Fork::isDone) || exceptionHandler.getException() != null).
         initialDelay(initialDelay).maxDelay(initialDelay).maxWait(TimeUnit.MINUTES.toMillis(maxWaitInMinute)).await()) {
       throw new TimeoutException("Forks did not finish withing specified timeout.");
     }
+    if (exceptionHandler.getException() != null) {
+      throw Throwables.propagate(exceptionHandler.getException());
+    }
```

Comment:
You mentioned a mock to test this functionality to fail the container and retry the job again, but I don't see the rests associated with that